vllm/platforms/cuda.py

```diff
@@ -118,29 +57,82 @@ def device_id_to_physical_device_id(device_id: int) -> int:
         return device_id
 
 
-class CudaPlatform(Platform):
+class CudaPlatformBase(Platform):
     _enum = PlatformEnum.CUDA
     device_type: str = "cuda"
     dispatch_key: str = "CUDA"
 
+    @classmethod
+    def get_device_capability(cls, device_id: int = 0) -> DeviceCapability:
+        raise NotImplementedError
+
+    @classmethod
+    def get_device_name(cls, device_id: int = 0) -> str:
+        raise NotImplementedError
+
+    @classmethod
+    def get_device_total_memory(cls, device_id: int = 0) -> int:
+        raise NotImplementedError
+
+    @classmethod
+    def is_full_nvlink(cls, device_ids: List[int]) -> bool:
+        raise NotImplementedError
+
+    @classmethod
+    def log_warnings(cls):
+        pass
+
+    @classmethod
+    def check_and_update_config(cls, vllm_config: VllmConfig) -> None:
+        parallel_config = vllm_config.parallel_config
+        scheduler_config = vllm_config.scheduler_config
+        if parallel_config.worker_cls == "auto":
+            if scheduler_config.is_multi_step:
+                parallel_config.worker_cls = \
+                    "vllm.worker.multi_step_worker.MultiStepWorker"
+            elif vllm_config.speculative_config:
+                parallel_config.worker_cls = \
+                    "vllm.spec_decode.spec_decode_worker.create_spec_worker"
+            else:
+                parallel_config.worker_cls = "vllm.worker.worker.Worker"
+
+
+# NVML utils
+# Note that NVML is not affected by `CUDA_VISIBLE_DEVICES`,
+# all the related functions work on real physical device ids.
+# the major benefit of using NVML is that it will not initialize CUDA
+class NvmlCudaPlatform(CudaPlatformBase):
+
+    @staticmethod
+    def with_nvml_context(fn: Callable[_P, _R]) -> Callable[_P, _R]:
+
+        @wraps(fn)
+        def wrapper(*args: _P.args, **kwargs: _P.kwargs) -> _R:
+            pynvml.nvmlInit()
+            try:
+                return fn(*args, **kwargs)
+            finally:
+                pynvml.nvmlShutdown()
+
+        return wrapper
```

Comment:
sorry for the additional comments. can you move this function outside of `NvmlCudaPlatform`?

Reply:
Done